File: android/src/main/java/com/jideguru/epub_viewer/EpubViewerPlugin.java
package com.jideguru.epub_viewer;

import android.app.Activity;
import android.content.Context;
import android.util.Log;

import java.util.Map;

import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import androidx.annotation.NonNull;

/** EpubReaderPlugin */
public class EpubViewerPlugin implements MethodCallHandler, FlutterPlugin, ActivityAware {

  private Reader reader;
  private ReaderConfig config;
  private MethodChannel channel;
  static private Activity activity;
  static private Context context;
  static BinaryMessenger messenger;
  static private EventChannel eventChannel;
  static private EventChannel.EventSink sink;
  private static final String channelName = "epub_viewer";

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {

    context = registrar.context();
    activity = registrar.activity();
    messenger = registrar.messenger();
    new EventChannel(messenger,"page").setStreamHandler(new EventChannel.StreamHandler() {

      @Override
      public void onListen(Object o, EventChannel.EventSink eventSink) {

        sink = eventSink;
        if(sink == null) {
          Log.i("empty", "Sink is empty");
        }
      }

      @Override
      public void onCancel(Object o) {

      }
    });


    final MethodChannel channel = new MethodChannel(registrar.messenger(), "epub_viewer");
    channel.setMethodCallHandler(new EpubViewerPlugin());

  }

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
    messenger = binding.getBinaryMessenger();
    context = binding.getApplicationContext();
    new EventChannel(messenger,"page").setStreamHandler(new EventChannel.StreamHandler() {

      @Override
      public void onListen(Object o, EventChannel.EventSink eventSink) {

        sink = eventSink;
        if(sink == null) {
          Log.i("empty", "Sink is empty");
        }
      }

      @Override
      public void onCancel(Object o) {

      }
    });
    channel = new MethodChannel(binding.getFlutterEngine().getDartExecutor(), channelName);
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    // TODO: your plugin is no longer attached to a Flutter experience.
  }

  @Override
  public void onAttachedToActivity(@NonNull ActivityPluginBinding activityPluginBinding) {
    activity = activityPluginBinding.getActivity();
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {

  }

  @Override
  public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding activityPluginBinding) {

  }

  @Override
  public void onDetachedFromActivity() {
    activity = null;
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {

    if (call.method.equals("setConfig")){
      Map<String,Object> arguments = (Map<String, Object>) call.arguments;
      String identifier = arguments.get("identifier").toString();
      String themeColor = arguments.get("themeColor").toString();
      String scrollDirection = arguments.get("scrollDirection").toString();
      Boolean nightMode = Boolean.parseBoolean(arguments.get("nightMode").toString());
      Boolean allowSharing = Boolean.parseBoolean(arguments.get("allowSharing").toString());
      Boolean enableTts = Boolean.parseBoolean(arguments.get("enableTts").toString());
      config = new ReaderConfig(context,identifier,themeColor,
              scrollDirection,allowSharing, enableTts,nightMode);

    } else if (call.method.equals("open")){

      Map<String,Object> arguments = (Map<String, Object>) call.arguments;
      String bookPath = arguments.get("bookPath").toString();
      String lastLocation = arguments.get("lastLocation").toString();

      Log.i("opening", "In open function");
      if(sink == null) {
        Log.i("sink status", "sink is empty");
      }
      reader = new Reader(context,messenger,config, sink);
      reader.open(bookPath, lastLocation);

    }else if(call.method.equals("close")){
      reader.close();
    }
    else if (call.method.equals("setChannel")){
      eventChannel = new EventChannel(messenger,"page");
      eventChannel.setStreamHandler(new EventChannel.StreamHandler() {

        @Override
        public void onListen(Object o, EventChannel.EventSink eventSink) {

          sink = eventSink;
        }

        @Override
        public void onCancel(Object o) {

        }
      });
    }

    else {
      result.notImplemented();
    }
  }
}

File: android/src/main/kotlin/com/jideguru/epub_viewer/EpubViewerPlugin.kt
package com.jideguru.epub_viewer

import android.app.Activity
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar
import timber.log.Timber
import kotlinx.coroutines.*
import kotlin.coroutines.*
import android.os.Looper
import android.content.Context
import android.os.AsyncTask
import android.os.Handler
import androidx.appcompat.app.AppCompatActivity
import android.content.Intent
import android.os.Bundle

/** EpubViewerPlugin */
class EpubViewerPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var activity: Activity
    private val mainScope = CoroutineScope(Dispatchers.Main)
    private var reader: Reader? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        reader = Reader()
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "epub_viewer")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        if (call.method.equals("setConfig")) {
            Timber.tag("epub_viewer").i("SetConfig")
        } else if (call.method.equals("open")) {
            val arguments = call.arguments as Map<String, Any>
            val bookPath = arguments["bookPath"].toString()
            val lastLocation = arguments["lastLocation"].toString()
            Timber.tag("epub_viewer").i("Open")


            val intent = Intent(context, Reader::class.java).apply {
                putExtra("bookPath", bookPath)
            }
            activity.startActivity(intent)
            // no thread
//            mainScope.launch {
//                try {
//                    withContext(Dispatchers.Main) {
//                        reader?.openBook(bookPath, lastLocation)
//                    }
//
//                    result.success(null)
//                } catch (e: Exception) {
//                    result.error("OpenBookException", e.message, null)
//                }
//            }
        } else if (call.method.equals("close")) {
//            reader.close()
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onDetachedFromActivity() {
        channel.setMethodCallHandler(null)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        /// TODO("Not yet implemented")
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity;
    }

    override fun onDetachedFromActivityForConfigChanges() {
        /// TODO("Not yet implemented")
    }
}
